### ISSUE TITLE
Http headers bust varnish cache server

### ISSUE THREAD

**Issue by pixeline** [wpninjas/ninja-forms#370]
I host my websites using Gandi.net's Simplehosting offer. It's a virtual machine coupled with a powerful varnish cache server. However, on a specific instance, the Varnish cache was not being used. After a lot of research, I desactivated all plugins (that made the Varnish server work), and reactivated them one by one until... activating ninja forms provoked the issue.

I looked at the ninja forms code and searched for header() calls, that might tell Varnish not to cache (such as Cache-Control, Pragma and Expires). Nothing seemed wrong there.

The problem may be related to the http headers, and the fact that you use php sessions (Wordpress core relies on clientside cookies).
I suggest using the php function session_cache_limiter(''); right before starting the session.
For instance: line 629 of ninja-forms.php could use that.

```` 
if( !session_id() ){
session_cache_limiter('');
session_start();
}
````
## right

Here is a "good" http header example (meaning the url will be served via the Varnish Cache):

```` 
$ curl -I domain.com/

HTTP/1.1 200 OK
Server: Apache/2.4.10
X-Powered-By: PHP/5.4.34-0+deb7u1
X-Pingback: http://domain.com/xmlrpc.php
Content-Type: text/html; charset=UTF-8
Vary: Accept-Encoding
Date: Sun, 02 Nov 2014 22:11:11 GMT
Connection: keep-alive
Via: 1.1 varnish
Age: 4
```` 

## wrong
Here is a "wrong" http header example, that is triggered when NinjaForms is activated. Notice the  "Set-Cookie" and the "Accept-Ranges" headers.

```` 
$ curl -I domain.com/

HTTP/1.1 200 OK
Server: Apache/2.4.10
X-Powered-By: PHP/5.4.34-0+deb7u1
X-Pingback: http://domain.com/xmlrpc.php
Set-Cookie: PHPSESSID=m5bvurqsv3a8gqt9j336nrdju1; path=/
Content-Type: text/html; charset=UTF-8
Accept-Ranges: bytes
Date: Sun, 02 Nov 2014 22:12:56 GMT
Connection: keep-alive
Via: 1.1 varnish
Age: 0
````

**Comment by kstover** [comment]
Sorry for the accidental close; I was cleaning up issues and got a little over-zealous.

I think that the best solution to this would be a session manager, as mentioned here: https://github.com/wpninjas/ninja-forms/issues/205.

**Comment by ottok** [comment]
How can I turn off sessions in Ninja Forms? It is a complete waste of resources and it bursts caches when Ninja Forms sets sessions on _every_ page of a site. It might make sense to start sessions on pages that actually display a form, but having a session for every user always at any page just does not make sense.

If you later make a session manager, the most important feature would be to allow site admins to disable the session features on at least the pages that don't have any Ninja Forms displayed.

**Comment by ottok** [comment]
I reviewed the code that sets the session. The session is set in this function:
```
	public function set_transient_id(){
		if( !session_id() )
	        session_start();
		if ( !isset ( $_SESSION['ninja_forms_transient_id'] ) AND !is_admin() ) {
			$t_id = ninja_forms_random_string();
			// Make sure that our transient ID isn't currently in use.
			while ( get_transient( $t_id ) !== false ) {
				$_id = ninja_forms_random_string();
			}
			$_SESSION['ninja_forms_transient_id'] = $t_id;
		}
	}
```

which is called on every single WP page load because when the plugin code is loaded it always runs this:
```
add_action( 'init', array( self::$instance, 'set_transient_id'), 1 );
```

The only place where this session is read that I can find is inside the ```ninja_forms_set_transient``` function:

```
	if ( isset ( $_SESSION['ninja_forms_transient_id'] ) ) {
		$transient_id = $_SESSION['ninja_forms_transient_id'];
	}
```

Here it is conditional and $transient_id might be stay undefined despite it is used a few lines below. The code is a bit buggy and to me it seems that the transient isn't actually used for anything, so that explains why nobody noticed the bugs - no functionality is broken. The set_transient function implies that form success and error messages are saved in a transient, which is seems against the idea of transients. If there is a message, the site can display it directly, or save in the session directly, but not save it in the database AND save a reference to the database entry in the session.


Eventually I added this to the theme functions.php to avoid Ninja Forms setting the unnecessary session and it seems to work fine:
```
/** 
 * Quickfix for Ninja Forms adding sessions on every page
 */

add_action( 'init', 'maybe_remove_ninja_forms_session', 0 );
function maybe_remove_ninja_forms_session() {
  $nf = Ninja_Forms();
  remove_action( 'init', array( $nf, 'set_transient_id'), 1 );
}
```

**Comment by pbogdan** [comment]
To also add to that there is one potential issue when disabling sessions this way. Looking through the code I see that `ninja_forms_set_transient()` is invoked only after form submissions if the form has been configured to do a redirect.

However with sessions disabled with the example Otto provided you will run into a problem with the following bit of code in `ninja_forms_set_transient()`:

```php
	if ( ! isset ( $_SESSION['ninja_forms_transient_id'] ) )
		ninja_forms_set_transient_id();
```

as the `ninja_forms_set_transient_id()` function isn't defined anywhere in the plugin so you will end up with a fatal error :-(. I guess it should be replaced with `Ninja_Forms::instance()->set_transient_id()` or fixed by defining a function with that single call as its body.

It would be great if at least that bit could be taken care of in the plugin so those who want to disable sessions can do so safely without having to play a game of `function_exists()` and creating our own version of the function.

Either way I think calling an undefined function would be worth fixing, regardless of how that relates to sessions, let me know and I can open a new issue for this if that makes more sense.

**Comment by kstover** [comment]
Hey. Thanks for the discussion! The biggest need for sessions at the moment is to keep track of user-submitted data between submission and redirect. For example, when using PayPal, the user is redirected to another site and back again. The submission data is then stored in a transient and the ID of that transient is in a session. It may make sense to move that to the specific extension that uses the feature until we have the time to create a better fix for sessions. I think it would also impact users who want a redirect that dislplays form data. 

That is probably a smaller subset of users, but it is a subset that does exist. 

A session manager, lik WP Sessions, would be the best solution, but we're up to our armpits internally, and I'm not sure when we are going to have time to implement it.

**Comment by kstover** [comment]
As of version 2.9.18, we've introduced an implementation of WP Session Manager, which means that we no longer use the PHP $_SESSION at all. This should take care of this issue.
